### PR TITLE
feat: x/leverage simulations

### PR DESCRIPTION
**Pull request by RafilxTenfen**
## Description

- Implemented operations functions of `x/leverage/simulation` module
  - Doing the tests we found an error related to the signature verification 
  - The Operations are NoOp for now
- Implemented decoder, params and genesis state with random params

ref: #354  

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/leverage/simulation/operations_test.go
Comment (toteki) [inline x/leverage/simulation/operations_test.go:29]:
```suggestion
// SetupTest creates a new umee base app
```

Comment (RafilxTenfen) [reply x/leverage/simulation/operations_test.go]: done at key := CreateCollateralAmountKeyNoDenom(lenderAddr)

**Comment by RafilxTenfen** [comment]
I had to add the Account Keeper to the leverage keeper to use the `simulation.GenAndDeliverTxWithRandFees(txCtx)` https://github.com/umee-network/umee/blob/3a78ce308f329f5115de44b1011882e2741fb11b/x/leverage/simulation/operations.go#L69 for just one func `GetAccount(ctx sdk.Context, addr sdk.AccAddress) types.AccountI` and only used in the simulations

- @toteki @alexanderbez

**Comment by toteki** [comment]
@alexanderbez Can you think of any way that we could get access to the `AccountKeeper` in leverage simulations without adding it to the leverage keeper?

@RafilxTenfen and I did not find a clean way to do that - though another option he came up with involved using the oracle keeper's  `AccountKeeper` in `leverage/simulation`.

File: x/leverage/keeper/keeper.go
Comment (toteki) [inline x/leverage/keeper/keeper.go:24]: I'm not sure we should be importing `cosmos-sdk/x/simulation` types in files outside `x/leverage/simulation` folder. We might need to find another way

Comment (RafilxTenfen) [reply x/leverage/keeper/keeper.go:24]: I could create an interface in the `x/leverage/types/expected_types.go` AccountKeeper

Comment (toteki) [reply x/leverage/keeper/keeper.go:24]: Bez sent me this to look at https://github.com/cosmos/cosmos-sdk/blob/master/x/distribution/simulation/operations.go#L27 and indicated `we need access to the account and bank keeper but that's only in the app module type, not the keeper type`.

If mimicking that, we should not need to modify `x/leverage/keeper` and `x/leverage/types` folders at all I think.

In `operations_test.go` of the `cosmos-sdk/x/distribution` module, the following appears for example:
```
op := simulation.SimulateMsgSetWithdrawAddress(suite.app.AccountKeeper, suite.app.BankKeeper, suite.app.DistrKeeper)
```
so the account keeper is being retrieved from suite.app

Comment (RafilxTenfen) [reply x/leverage/keeper/keeper.go:24]: To check if I got it correctly it should add the AccountKeeper to the leverage `AppModule` structure instead of the `Keeper` 
Just like it happen here? https://github.com/cosmos/cosmos-sdk/blob/b5ca8e0b48279a74e126721929764d1ef09d3449/x/distribution/module.go#L204-L205
@toteki

Comment (alexanderbez) [reply x/leverage/keeper/keeper.go:24]: Yes, correct. There should be no reason to modify `x/leverage`'s `Keeper` type. Any external keeper dependencies needed for simulations should be added to the `AppModule` directly 👍

File: x/leverage/keeper/keeper.go
Comment (toteki) [inline x/leverage/keeper/keeper.go:610]: This can be moved to `app.go` with the signature

`func (am AppModule) WeightedOperations(simState module.SimulationState) []simtypes.WeightedOperation`

similar to how it's done [here](https://github.com/cosmos/cosmos-sdk/blob/master/x/distribution/module.go#L203), so keeper doesn't need to import simulation stuff

**Review by toteki — CHANGES_REQUESTED** [review]

File: x/leverage/keeper/keeper.go
Comment (RafilxTenfen) [reply x/leverage/keeper/keeper.go:610]: I will do that, this way we need to add the `bankKeeper` to the  `AppModule` structure as well

Comment (RafilxTenfen) [reply x/leverage/keeper/keeper.go]: done at https://github.com/umee-network/umee/pull/401/commits/20cd42a30a344993c368a6c33163cc1a66b5d7b4

File: x/leverage/keeper/keeper.go
Comment (toteki) [inline x/leverage/keeper/keeper.go:12]:
```suggestion

	"github.com/umee-network/umee/x/leverage/types"
```
apply import conventions

File: x/leverage/module.go
Comment (toteki) [inline x/leverage/module.go:18]:
```suggestion

	"github.com/umee-network/umee/x/leverage/client/cli"
```

File: x/leverage/simulation/operations.go
Comment (toteki) [inline x/leverage/simulation/operations.go:12]:
```suggestion

	"github.com/umee-network/umee/x/leverage/types"
```

File: x/leverage/simulation/operations_test.go
Comment (toteki) [inline x/leverage/simulation/operations_test.go:18]:
```suggestion
	sdk "github.com/cosmos/cosmos-sdk/types"
	simtypes "github.com/cosmos/cosmos-sdk/types/simulation"
	minttypes "github.com/cosmos/cosmos-sdk/x/mint/types"
	"github.com/stretchr/testify/suite"
	tmrand "github.com/tendermint/tendermint/libs/rand"
	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"

	umeeapp "github.com/umee-network/umee/app"
	umeeappbeta "github.com/umee-network/umee/app/beta"
	"github.com/umee-network/umee/x/leverage/simulation"
	"github.com/umee-network/umee/x/leverage/types"
```

File: x/leverage/keeper/keeper.go
Comment (RafilxTenfen) [reply x/leverage/keeper/keeper.go]: done at

File: x/leverage/simulation/operations.go
Comment (RafilxTenfen) [reply x/leverage/simulation/operations.go:13]: done at https://github.com/umee-network/umee/pull/401/commits/9b1ecfda4c9bfbdfc04c43cdd6dd4b82bf54f279

File: x/leverage/simulation/operations_test.go
Comment (RafilxTenfen) [reply x/leverage/simulation/operations_test.go]: done at https://github.com/umee-network/umee/pull/401/commits/9b1ecfda4c9bfbdfc04c43cdd6dd4b82bf54f279

File: x/leverage/simulation/genesis.go
Comment (toteki) [inline x/leverage/simulation/genesis.go:23]:
```suggestion
	return int64(10 + r.Intn(90))
```
For range to end at 100

File: x/leverage/types/expected_types.go
Comment (toteki) [inline x/leverage/types/expected_types.go:5]:
```suggestion
	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
```

File: x/leverage/types/expected_types.go
Comment (toteki) [inline x/leverage/types/expected_types.go:10]:
```suggestion
	GetAccount(ctx sdk.Context, addr sdk.AccAddress) authtypes.AccountI
```

File: x/leverage/types/expected_types.go
Comment (toteki) [inline x/leverage/types/expected_types.go:29]:
```suggestion
```
Currently not used

File: x/leverage/simulation/decoder.go
Comment (toteki) [inline x/leverage/simulation/decoder.go:56]: `LastInterestTime` is now being stored as an int64 using `gogotypes.Int64Value` (instead of `sdk.Int`) as seen in `x/leverage/keeper/interest.go` near the bottom of the file. This function should be updated to do the same

**Review by toteki — APPROVED** [review]
Approving now. I would suggest

- changing PR description to indicate that operations are NoOp for now
- waiting for feedback from @alexanderbez before merging

File: x/leverage/simulation/decoder.go
Comment (RafilxTenfen) [reply x/leverage/simulation/decoder.go]: done at https://github.com/umee-network/umee/pull/401/commits/31a811a288bae32cc0fdf556720a5ea6df9128ad

File: x/leverage/types/expected_types.go
Comment (alexanderbez) [inline x/leverage/types/expected_types.go:9]: @RafilxTenfen @toteki do we actually need this or is this left over code from previous attempts?

**Pull request merged**
